### ISSUE TITLE
HL7 deserialization is incorrect when subcomponents are used

### ISSUE THREAD

**Issue by rbeckman-nextgen** [nextgenhealthcare/connect#3096]
It appears that right now, subcomponent deserialization is not working in various cases.

Try this out:
\\
\\
`msg.MSH['MSH.3']['MSH.3.1']['MSH.3.1.1'] = 'a';
msg.MSH['MSH.3']['MSH.3.1']['MSH.3.1.2'] = 'b';`
\\
As expected, the encoded data is "MSH](^~\&|a&b". However, now add the following line:
\\
\\
`msg.MSH['MSH.4')['MSH.4.1']['MSH.4.1.4'] = 'c';`
\\
Now the encoded data is "MSH](^~\&|a&b|&c", which is incorrect. The transformed data is correct though:
\\
\\
`<HL7Message>
    <MSH>
        <MSH.1>|</MSH.1>
        <MSH.2>^~\&amp;</MSH.2>
        <MSH.3>
            <MSH.3.1>
                <MSH.3.1.1>a</MSH.3.1.1>
                <MSH.3.1.2>b</MSH.3.1.2>
            </MSH.3.1>
        </MSH.3>
        <MSH.4>
            <MSH.4.1>
                <MSH.4.1.4>c</MSH.4.1.4>
            </MSH.4.1>
        </MSH.4>
    </MSH>
</HL7Message>`
\\
This is because we store an array in memory in the SAX handler that indicates the _last_ subcomponent parsed (e.g. [MSH, 3, 1, 2)). We do the same thing at the field and component levels, but obviously when a new segment or field is encountered, that array is reset. For whatever reason, this is just not being done for the subcomponent level. Depending on what the last subcomponent ID was, this can cause delimiters to be omitted incorrectly throughout the message.

The actual changes are simple, basically two lines. I'll attach patches for 2.x/3.x.

Imported Issue. Original Details:
Jira Issue Key: MIRTH-3184
Reporter: narupley
Created: 2014-03-13T08:56:16.000-0700

**Comment by rbeckman-nextgen** [comment]
This actually also occurs when you have repeating fields, and the first _component_ in a subsequent field isn't present. This is because we're not resetting the _component_ name array after adding the repetition marker in the SAX handler. I'll update the patches.

Imported Comment. Original Details:
Author: narupley
Created: 2014-03-17T10:39:17.000-0700

**Comment by rbeckman-nextgen** [comment]
Fixed in revisions 7255/7256.

Imported Comment. Original Details:
Author: narupley
Created: 2014-08-01T12:31:25.000-0700

**Comment by rbeckman-nextgen** [comment]
Verified the issue existed in older builds by creating a channel, adding a transformer step to add in various subcomponents and blasting it with sample hl7 messages. The channel was exported to 3.1 where I transmitted more sample hl7 messages. The contents of the transformed and encoded data were inspected to ensure deserialization is handled correctly and that nothing else was inadvertently affected.

Imported Comment. Original Details:
Author: jaysenp
Created: 2014-08-20T09:51:35.000-0700
